compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,418 @@
+            // u3dynI<length><name>E as vendor extended type
+            let mut s = String::from("u3dynI");
+            for predicate in predicates.iter() {
+                match predicate.as_ref().skip_binder() {
+                    ty::ExistentialPredicate::Trait(trait_ref) => {
+                        let name = format!("{}", tcx.opt_item_name(trait_ref.def_id)?);
+                        s.push_str(&format!("{}", name.len()));
+                        s.push_str(&name);
+                    }
+                    ty::ExistentialPredicate::Projection(_) => {}
+                    ty::ExistentialPredicate::AutoTrait(_) => {}
+                };
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Placeholder types
+        ty::Bound(..) | ty::Error(_) | ty::Infer(..) | ty::Param(_) | ty::Placeholder(_) => {
+            // u1p as vendor extended type
+            let mut s = String::from("u1p");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::GeneratorWitness(..) => {
+            bug!("encode_ty: unexpected `GeneratorWitness`");
+        }
+
+        ty::Opaque(..) => {
+            bug!("encode_ty: unexpected `Opaque`");
+        }
+
+        ty::Projection(..) => {
+            bug!("encode_ty: unexpected `Projection`");
+        }
+    };
+
+    Some(typeid.to_string())
+}
+
+/// Returns a type metadata identifier for the specified FnAbi using the Itanium C++ ABI with vendor
+/// extended type qualifiers and types for Rust types that are not used at the FFI boundary.
+pub fn typeid_for_fnabi<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> Option<String> {
+    // A name is mangled by prefixing "_Z" to an encoding of its name, and in the case of functions
+    // its type.
+    let mut typeid = String::from("_Z");
+
+    // Clang uses the Itanium C++ ABI's virtual tables and RTTI typeinfo structure name as type
+    // metadata identifiers for function pointers. The typeinfo name encoding is a two-character
+    // code (i.e., 'TS') prefixed to the type encoding for the function.
+    typeid.push_str("TS");
+
+    // Function types are delimited by an "F..E" pair
+    typeid.push('F');
+
+    // A dictionary of substitution candidates used for compression (see
+    // https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-compression).
+    let mut dict: FxHashMap<(Ty<'tcx>, TyQ), usize> = FxHashMap::default();
+
+    // Identify C char and integer type uses and encode them with their respective builtin type
+    // encoding as specified by the Itanium C++ ABI for extern function types with the "C" calling
+    // convention to use this encoding for cross-language LLVM CFI.
+    let encode_c_types = match fn_abi.conv {
+        // FIXME(rcvalle): See comment before the ty::Adt match in encode_ty().
+        Conv::C => false,
+        _ => false,
```

Comment:
Maybe use the rust v0 symbol mangling scheme for the Rust and rust-call calling conventions? That would also ensure that we don't call a function with the Rust call conv using the C call conv.

Reply:
See comment above. (Adding the calling convention to the encoding may be something that we will do in the future.)